src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -1275,7 +1279,150 @@ void mongoSubCacheRefresh(void)
   //
 #endif
 
-  cacheSemGive(__FUNCTION__, "subscription cache for refresh");
+  if (!semAlreadyTaken)
+  {
+    cacheSemGive(__FUNCTION__, "subscription cache for refresh");
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* CachedSubSaved - 
+*/
+typedef struct CachedSubSaved
+{
+  long long  lastNotificationTime;
+  long long  count;
+} CachedSubSaved;
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheSync - 
+*
+* 1. Save subscriptionId, lastNotificationTime, and count for all items in cache (savedSubV)
+* 2. Refresh cache (count set to 0)
+* 3. Compare lastNotificationTime in savedSubV with the new cache-contents and:
+*    3.1 Update cache-items where 'saved lastNotificationTime' > 'cached lastNotificationTime'
+*    3.2 Remember this more correct lastNotificationTime (must be flushed to mongo) -
+*        by clearing out (set to 0) those lastNotificationTimes that are newer in cache
+* 4. Update 'count' for each item in savedSubV where count != 0
+* 5. Update 'lastNotificationTime' foreach item in savedSubV where lastNotificationTime != 0
+* 6. Free the vector created in step 1 - savedSubV
+*/
+void mongoSubCacheSync(void)
+{
+  std::map<std::string, CachedSubSaved*> savedSubV;
+
+  cacheSemTake(__FUNCTION__, "Synchronizing subscription cache");
+
+
+  //
+  // 1. Save subscriptionId, lastNotificationTime, and count for all items in cache
+  //
+  CachedSubscription* cSubP = mongoSubCache.head;
+
+  while (cSubP != NULL)
+  {
+    CachedSubSaved* cssP       = new CachedSubSaved();
+
+    cssP->lastNotificationTime = cSubP->lastNotificationTime;
+    cssP->count                = cSubP->count;
+
+    savedSubV[cSubP->subscriptionId]= cssP;
+    cSubP = cSubP->next;
+  }
+
+  LM_T(LmtMongoCacheSync, ("Pushed back %d items to savedSubV", savedSubV.size()));
+
+
+  //
+  // 2. Refresh cache (count set to 0)
+  //
+  mongoSubCacheRefresh(true);
```

Comment:
Maybe the semAlreadyTaken parameter is not really needed, as I understand the only method that call mongoSubCacheRefresh() is mongoSubCacheSync() (except during initialization, which is a moment in which semaphores are not needed.

Thus, mongoSubCacheRefresh() could be simplified to not use this parameter. Regarding the effective, cacheSemTake/cacheSemGive, not sure if the best place is inside mongoSubCacheRefresh() or as a "ward" when invoking mongoSubCacheRefresh, i.e. something like this at this pooint of code:

```
cacheSemTake(...);
mongoSubCacheRefresh();
cacheSemGive(...);
```

Inspired in other places, we we do:

```
   reqSemTake(__FUNCTION__, "mongoSubCacheStatisticsGet", SemReadOp, &reqSemTaken);
   mongoSubCacheStatisticsGet(&mscRefreshs, &mscInserts, &mscRemoves, &mscUpdates, &cacheItems, listBuffer, sizeof(listBuffer));
   reqSemGive(__FUNCTION__, "mongoSubCacheStatisticsGet", reqSemTaken);
```

(In that place I think the sem is wrong, but the idea is the same).